src/hotspot/share/cds/cdsConstants.cpp

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2014, 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include <cstddef>
+#include "cds.h"
+#include "cds/cdsConstants.hpp"
+#include "cds/dynamicArchive.hpp"
+#include "cds/filemap.hpp"
+
+struct {
+  const char* _name;
+  size_t _size;
+} cds_offsets[] = {
+  { "CDSFileMapHeaderBase::_magic",           offset_of(CDSFileMapHeaderBase, _magic)           },
+  { "CDSFileMapHeaderBase::_crc",             offset_of(CDSFileMapHeaderBase, _crc)             },
+  { "CDSFileMapHeaderBase::_version",         offset_of(CDSFileMapHeaderBase, _version)         },
+  { "CDSFileMapHeaderBase::_space[0]",        offset_of(CDSFileMapHeaderBase, _space)           },
+  { "FileMapHeader::_jvm_ident",              offset_of(FileMapHeader, _jvm_ident)              },
+  { "FileMapHeader::_base_archive_name_size", offset_of(FileMapHeader, _base_archive_name_size) },
+  { "CDSFileMapRegion::_crc",                 offset_of(CDSFileMapRegion, _crc)                 },
+  { "CDSFileMapRegion::_used",                offset_of(CDSFileMapRegion, _used)                },
+  { "DynamicArchiveHeader::_base_region_crc", offset_of(DynamicArchiveHeader, _base_region_crc) }
+};
+
+constexpr struct {
+  const char* _name;
+  int         _value;
+} cds_constants[] = {
+  { "static_magic",                 (int)CDS_ARCHIVE_MAGIC         },
+  { "dynamic_magic",                (int)CDS_DYNAMIC_ARCHIVE_MAGIC },
+  { "int_size",                     sizeof(int)                    },
+  { "CDSFileMapRegion_size",        sizeof(CDSFileMapRegion)       },
+  { "static_file_header_size",      sizeof(FileMapHeader)          },
+  { "dynamic_archive_header_size",  sizeof(DynamicArchiveHeader)   },
+  { "size_t_size",                  sizeof(size_t)                 }
+};
+
+size_t get_cds_offset(const char* name) {
+  for (int i = 0; i < (int)(sizeof(cds_offsets)/sizeof(cds_offsets[0])); i++) {
```

Comment:
Can't we just define a size for the array and use that? It isn't really a dynamic/unknown quantity.

Reply:
The array size is known at compile time --- it is not dynamic but we do not need to manually count the size of the array this way. If define a size for the array, we need to count the number of items in the array to define the array size. We may add more items to the array --- just add items and don't need change anything else. Since total array memory size and array component size are known at compile time, so compiler will not generate code to calculate the size. How about this? 
const int size =   (int)(sizeof(cds_offsets)/sizeof(cds_offsets[0]));
for (int i = 0; i < size; i++) {
  ...
}

Reply:
I always thought that was a hack way of determining an array size, not an approved method. :(

Don't worry just leave it as-is.

Reply:
We have functionality for this in globalDefinitions.hpp:
```
#define ARRAY_SIZE(array) sizeof(array_size_impl(array))
```

Reply:
@stefank Thanks --- it saves code size.